sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/evaluate/ResourceEvaluationStage.java

```diff
@@ -0,0 +1,189 @@
+                Value unreserveValue = ValueUtils.subtract(mesosResource.getValue(), resourceRequirement.getValue());
+
+                if (ValueUtils.compare(unreserveValue, ValueUtils.getZero(unreserveValue.getType())) > 0) {
+                    logger.info("Updates reserved resource with less reservation");
+                    Resource unreserveResource = ResourceUtils.getDesiredResource(
+                            resourceRequirement.getRole(),
+                            resourceRequirement.getPrincipal(),
+                            resourceRequirement.getName(),
+                            unreserveValue);
+                    unreserveResource = ResourceUtils.setResourceId(
+                            unreserveResource, resourceRequirement.getResourceId());
+
+                    offerResourcePool.release(new MesosResource(
+                            ResourceUtils.getUnreservedResource(resourceRequirement.getName(), unreserveValue)));
+                    offerRecommendationSlate.addUnreserveRecommendation(
+                            new UnreserveOfferRecommendation(offerResourcePool.getOffer(), unreserveResource));
+                    fulfilledResource = getFulfilledResource(new MesosResource(resourceRequirement.getResource()));
+                }
+
+                if (ValueUtils.compare(reserveValue, ValueUtils.getZero(reserveValue.getType())) > 0) {
+                    logger.info("Updates reserved resource with additional reservation");
+                    Resource reserveResource = ResourceUtils.getDesiredResource(
+                            resourceRequirement.getRole(),
+                            resourceRequirement.getPrincipal(),
+                            resourceRequirement.getName(),
+                            reserveValue);
+
+                    if (offerResourcePool.consume(new ResourceRequirement(reserveResource)).isPresent()) {
+                        reserveResource = ResourceUtils.setResourceId(
+                                reserveResource, resourceRequirement.getResourceId());
+                        offerRecommendationSlate.addReserveRecommendation(
+                                new ReserveOfferRecommendation(offerResourcePool.getOffer(), reserveResource));
+                        fulfilledResource = getFulfilledResource(new MesosResource(resourceRequirement.getResource()));
+                    } else {
+                        throw new OfferEvaluationException("Insufficient resources to increase resource usage.");
+                    }
+                }
+            }
+        } else {
+            if (resourceRequirement.reservesResource()) {
+                logger.info("Reserves Resource");
+                offerRecommendationSlate.addReserveRecommendation(
+                        new ReserveOfferRecommendation(offerResourcePool.getOffer(), fulfilledResource));
+            }
+
+            if (resourceRequirement.createsVolume()) {
```

Comment:
It could -- I didn't put it there initially because the super class's `evaluate` already consumes the resource, and so it only has one chance to make decisions based on the `MesosResource` that the pool returns, so putting this logic in `VolumeEvaluationStage` would entail duplicating that code, but of course almost none of the code in `ResourceEvaluationStage` applies since it's mostly dealing with bookkeeping for non-atomic resources. I'll move this.